app/src/editor/canvas/tests/canvas.test.js

```diff
@@ -0,0 +1,57 @@
+import * as State from '../state'
+import * as Mocks from './mocks'
+
+describe('Canvas Module', () => {
+    it('has a port to select canvas', () => {
+        let canvas = State.emptyCanvas()
+        canvas = State.addModule(canvas, Mocks.Canvas())
+
+        const subCanvas = canvas.modules.find((m) => m.name === 'Canvas')
+        expect(subCanvas).toBeTruthy()
+
+        const canvasPort = State.findModulePort(canvas, subCanvas.hash, (p) => p.name === 'canvas')
+        expect(canvasPort).toBeTruthy()
+    })
+
+    it('reloads inputs and outputs from selected canvas', () => {
+        let canvas = State.emptyCanvas()
+        canvas = State.addModule(canvas, Mocks.Canvas())
+        const { hash } = canvas.modules.find((m) => m.jsModule === 'CanvasModule')
+
+        const newModule = Mocks.CanvasWithSelected()
+        const newCanvas = State.updateCanvas(State.updateModule(canvas, hash, () => newModule))
+        const subCanvas = newCanvas.modules.find((m) => m.jsModule === 'CanvasModule')
+
+        expect(subCanvas.params.length === newModule.params.length).toBeTruthy()
```

Comment:
`expect(subCanvas.params.length).toBe(newModule.params.length)` will produce a more revealing error on failure e.g.  "expected: true, received: false" doesn't indicate what the difference is, whether one module has zero, which module had an issue etc. Compare with  "expected: 0, received: 1". This gives some indication as to which module  might be having an issue and the nature of the failure. e.g. no data vs off-by-one